File: test/integration.test.js
const http = require('http');

// Test configuration
const TEST_HOST = process.env.TEST_HOST || 'localhost';
const TEST_PORT = process.env.TEST_PORT || 8070;
const TEST_TIMEOUT = 30000; // 30 seconds

// Test coordinates (Tallinn, Estonia)
const TEST_LAT = 59.437;
const TEST_LON = 24.7536;

function makeGraphQLRequest(query, variables = {}) {
  return new Promise((resolve, reject) => {
    const postData = JSON.stringify({
      query,
      variables
    });

    const options = {
      hostname: TEST_HOST,
      port: TEST_PORT,
      path: '/graphql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData)
      },
      timeout: TEST_TIMEOUT
    };

    const req = http.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          resolve({ statusCode: res.statusCode, body: parsed });
        } catch (e) {
          reject(new Error(`Failed to parse response: ${e.message}`));
        }
      });
    });

    req.on('error', (e) => {
      reject(new Error(`Request failed: ${e.message}`));
    });

    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });

    req.write(postData);
    req.end();
  });
}

async function testHealthCheck() {
  return new Promise((resolve, reject) => {
    const req = http.request({
      hostname: TEST_HOST,
      port: TEST_PORT,
      path: '/health',
      method: 'GET',
      timeout: 5000
    }, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try {
          const parsed = JSON.parse(data);
          if (parsed.hello === 'world') {
            console.log('✓ Health check passed');
            resolve(true);
          } else {
            reject(new Error('Health check failed: unexpected response'));
          }
        } catch (e) {
          reject(new Error(`Health check failed: ${e.message}`));
        }
      });
    });

    req.on('error', reject);
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Health check timeout'));
    });
    req.end();
  });
}

async function testWeatherQuery() {
  const query = `
    query GetWeather($lat: String!, $lng: String!) {
      weather(lat: $lat, lng: $lng)
    }
  `;

  const response = await makeGraphQLRequest(query, {
    lat: TEST_LAT.toString(),
    lng: TEST_LON.toString()
  });

  if (response.statusCode !== 200) {
    throw new Error(`Expected status 200, got ${response.statusCode}`);
  }

  if (response.body.errors) {
    throw new Error(`GraphQL errors: ${JSON.stringify(response.body.errors)}`);
  }

  const weather = response.body.data?.weather;
  if (!weather) {
    throw new Error('No weather data returned');
  }

  // Weather returns JSON object
  console.log('✓ Weather query returned data:', JSON.stringify(weather).substring(0, 200) + '...');

  return true;
}

async function testForecastQuery() {
  const query = `
    query GetHistoricalWeather($lat: String!, $lng: String!, $startDate: String!, $endDate: String!) {
      historicalWeather(lat: $lat, lng: $lng, startDate: $startDate, endDate: $endDate) {
        temperature {
          temperature_2m {
            time
            value
          }
        }
      }
    }
  `;

  // Get dates for the last 7 days
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 7);

  const response = await makeGraphQLRequest(query, {
    lat: TEST_LAT.toString(),
    lng: TEST_LON.toString(),
    startDate: startDate.toISOString().split('T')[0],
    endDate: endDate.toISOString().split('T')[0]
  });

  if (response.statusCode !== 200) {
    throw new Error(`Expected status 200, got ${response.statusCode}`);
  }

  if (response.body.errors) {
    throw new Error(`GraphQL errors: ${JSON.stringify(response.body.errors)}`);
  }

  const historical = response.body.data?.historicalWeather;
  if (!historical) {
    throw new Error('No historical weather data returned');
  }

  const tempData = historical.temperature?.temperature_2m;
  if (!tempData || !Array.isArray(tempData) || tempData.length === 0) {
    throw new Error('No temperature data in historical weather');
  }

  console.log(`✓ Historical weather query returned ${tempData.length} temperature readings`);
  
  return true;
}

async function runTests() {
  console.log('Starting integration tests...\n');
  
  const tests = [
    { name: 'Health Check', fn: testHealthCheck },
    { name: 'Weather Query', fn: testWeatherQuery },
    { name: 'Historical Weather Query', fn: testForecastQuery }
  ];

  let passed = 0;
  let failed = 0;

  for (const test of tests) {
    try {
      console.log(`Running: ${test.name}`);
      await test.fn();
      passed++;
    } catch (error) {
      console.error(`✗ ${test.name} failed:`, error.message);
      failed++;
    }
    console.log('');
  }

  console.log('─'.repeat(50));
  console.log(`Tests completed: ${passed} passed, ${failed} failed`);
  console.log('─'.repeat(50));

  if (failed > 0) {
    process.exit(1);
  }

  process.exit(0);
}

// Run tests
runTests().catch((error) => {
  console.error('Test runner failed:', error);
  process.exit(1);
});

File: integration_test.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	logger "github.com/Gratheon/log-lib-go"
)

const (
	testLat = "59.437"
	testLng = "24.7536"
)

func TestHealthCheck(t *testing.T) {
	server := newIntegrationServer(t)
	defer server.Close()

	resp, err := server.Client().Get(server.URL + "/health")
	if err != nil {
		t.Fatalf("health request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("expected status 200, got %d", resp.StatusCode)
	}

	var body map[string]string
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		t.Fatalf("decode health response: %v", err)
	}

	if body["hello"] != "world" {
		t.Fatalf("unexpected health payload: %#v", body)
	}
}

func TestWeatherQuery(t *testing.T) {
	server := newIntegrationServer(t)
	defer server.Close()

	response := graphqlRequest(t, server, `
		query GetWeather($lat: String!, $lng: String!) {
			weather(lat: $lat, lng: $lng)
		}
	`, map[string]string{
		"lat": testLat,
		"lng": testLng,
	})

	weather, ok := response.Data["weather"].(map[string]interface{})
	if !ok || len(weather) == 0 {
		t.Fatalf("expected weather object, got %#v", response.Data["weather"])
	}
}

func TestHistoricalWeatherQuery(t *testing.T) {
	server := newIntegrationServer(t)
	defer server.Close()

	endDate := time.Now().UTC()
	startDate := endDate.AddDate(0, 0, -7)

	response := graphqlRequest(t, server, `
		query GetHistoricalWeather($lat: String!, $lng: String!, $startDate: String!, $endDate: String!) {
			historicalWeather(lat: $lat, lng: $lng, startDate: $startDate, endDate: $endDate) {
				temperature {
					temperature_2m {
						time
						value
					}
				}
			}
		}
	`, map[string]string{
		"lat":       testLat,
		"lng":       testLng,
		"startDate": startDate.Format("2006-01-02"),
		"endDate":   endDate.Format("2006-01-02"),
	})

	historical, ok := response.Data["historicalWeather"].(map[string]interface{})
	if !ok {
		t.Fatalf("expected historicalWeather object, got %#v", response.Data["historicalWeather"])
	}

	temperature, ok := historical["temperature"].(map[string]interface{})
	if !ok {
		t.Fatalf("expected temperature object, got %#v", historical["temperature"])
	}

	readings, ok := temperature["temperature_2m"].([]interface{})
	if !ok || len(readings) == 0 {
		t.Fatalf("expected temperature readings, got %#v", temperature["temperature_2m"])
	}
}

func newIntegrationServer(t *testing.T) *httptest.Server {
	t.Helper()

	logger.Configure(logger.LoggerConfig{
		LogLevel: logger.LogLevelError,
	})

	svc, err := newService(config{
		Port:                             0,
		SchemaRegistryHost:               "http://127.0.0.1:0",
		SelfURL:                          "weather:8070",
		RedisHost:                        "",
		RedisPort:                        6379,
		RedisPassword:                    "",
		RedisDB:                          0,
		HistoricalWeatherCacheTTLSeconds: 1800,
		EnvironmentID:                    "test",
		LogLevel:                         "error",
	})
	if err != nil {
		t.Fatalf("create service: %v", err)
	}

	mux := http.NewServeMux()
	mux.HandleFunc("/", rootHandler)
	mux.HandleFunc("/health", healthHandler)
	mux.Handle("/graphql", svc.graphqlHandler())

	return httptest.NewServer(loggingMiddleware(mux))
}

type graphqlResponse struct {
	Data   map[string]interface{} `json:"data"`
	Errors []interface{}          `json:"errors"`
}

func graphqlRequest(t *testing.T, server *httptest.Server, query string, variables map[string]string) graphqlResponse {
	t.Helper()

	payload, err := json.Marshal(map[string]interface{}{
		"query":     query,
		"variables": variables,
	})
	if err != nil {
		t.Fatalf("marshal graphql payload: %v", err)
	}

	req, err := http.NewRequestWithContext(context.Background(), http.MethodPost, server.URL+"/graphql", bytes.NewReader(payload))
	if err != nil {
		t.Fatalf("build graphql request: %v", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := server.Client().Do(req)
	if err != nil {
		t.Fatalf("graphql request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		t.Fatalf("expected status 200, got %d: %s", resp.StatusCode, string(body))
	}

	var parsed graphqlResponse
	if err := json.NewDecoder(resp.Body).Decode(&parsed); err != nil {
		t.Fatalf("decode graphql response: %v", err)
	}

	if len(parsed.Errors) > 0 {
		t.Fatalf("graphql errors: %s", formatJSON(parsed.Errors))
	}

	return parsed
}

func formatJSON(value interface{}) string {
	raw, err := json.Marshal(value)
	if err != nil {
		return fmt.Sprintf("%#v", value)
	}

	return string(raw)
}
